File: api/schemas/model_config.py
"""API schemas for workspace BYOK model configuration."""
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field


class CredentialStatusResponse(BaseModel):
    provider: str = "openrouter"
    configured: bool = False
    status: str = ""
    masked_key: str = ""
    last_validated_at: str | None = None
    last_validation_error: str = ""


class AgentCatalogItemResponse(BaseModel):
    agent_key: str = ""
    display_name: str = ""
    group: str = ""
    pipeline: str = ""
    pipeline_step: str = ""
    default_model: str = ""
    capabilities: list[str] = Field(default_factory=list)
    required: bool = True
    description: str = ""


class AgentModelConfigResponse(BaseModel):
    agent_key: str = ""
    model: str = ""
    provider: str = "openrouter"
    temperature: float | None = None
    max_tokens: int | None = None
    timeout_s: float | None = None
    enabled: bool = True
    uses_default: bool = True
    updated_at: str | None = None


class WorkspaceModelConfigResponse(BaseModel):
    workspace_slug: str = ""
    credential: CredentialStatusResponse = Field(default_factory=CredentialStatusResponse)
    catalog: list[AgentCatalogItemResponse] = Field(default_factory=list)
    configs: list[AgentModelConfigResponse] = Field(default_factory=list)
    missing_required_agent_keys: list[str] = Field(default_factory=list)


class OpenRouterKeyUpsertRequest(BaseModel):
    api_key: str = ""


class OpenRouterKeyTestRequest(BaseModel):
    api_key: str | None = None


class AgentModelConfigUpdateRequest(BaseModel):
    model: str = ""
    temperature: float | None = None
    max_tokens: int | None = None
    timeout_s: float | None = None
    enabled: bool | None = None
    extra_body: dict[str, Any] | None = None


class ModelConfigPreflightRequest(BaseModel):
    agent_keys: list[str] = Field(default_factory=list)


class CredentialTestResponse(BaseModel):
    ok: bool = False
    provider: str = "openrouter"
    model: str = ""
    error: str = ""


class AgentConfigTestResponse(BaseModel):
    ok: bool = False
    agent_key: str = ""
    model: str = ""
    error: str = ""


class ModelConfigPreflightResponse(BaseModel):
    ok: bool = False
    missing_credential: bool = False
    invalid_credential: bool = False
    missing_agent_keys: list[str] = Field(default_factory=list)
    disabled_agent_keys: list[str] = Field(default_factory=list)
    errors: list[str] = Field(default_factory=list)


class DeleteOpenRouterKeyResponse(BaseModel):
    deleted: bool = False

File: frontend/src/lib/model-config.ts
import { api, workspaceQueryParams } from '@/lib/api-client';

export interface CredentialStatusAPI {
  provider: string;
  configured: boolean;
  status: string;
  masked_key: string;
  last_validated_at: string | null;
  last_validation_error: string;
}

export interface AgentCatalogItemAPI {
  agent_key: string;
  display_name: string;
  group: string;
  pipeline: string;
  pipeline_step: string;
  default_model: string;
  capabilities: string[];
  required: boolean;
  description: string;
}

export interface AgentModelConfigAPI {
  agent_key: string;
  model: string;
  provider: string;
  temperature: number | null;
  max_tokens: number | null;
  timeout_s: number | null;
  enabled: boolean;
  uses_default: boolean;
  updated_at: string | null;
}

export interface AgentUsageSummaryAPI {
  agent_key: string;
  call_count: number;
  prompt_tokens: number;
  completion_tokens: number;
  estimated_cost_usd: number;
  last_used_at: string | null;
}

export interface WorkspaceModelConfigAPI {
  workspace_slug: string;
  credential: CredentialStatusAPI;
  catalog: AgentCatalogItemAPI[];
  configs: AgentModelConfigAPI[];
  missing_required_agent_keys: string[];
  usage_summary: AgentUsageSummaryAPI[];
}

export interface OpenRouterKeyUpsertRequestAPI {
  api_key: string;
}

export interface OpenRouterKeyTestRequestAPI {
  api_key?: string | null;
}

export interface CredentialTestResponseAPI {
  ok: boolean;
  provider: string;
  model: string;
  error: string;
}

export interface AgentModelConfigUpdateRequestAPI {
  model: string;
  temperature: number | null;
  max_tokens: number | null;
  timeout_s: number | null;
  enabled?: boolean | null;
}

export interface AgentConfigTestResponseAPI {
  ok: boolean;
  agent_key: string;
  model: string;
  error: string;
}

export interface DeleteOpenRouterKeyResponseAPI {
  deleted: boolean;
}

export interface ModelConfigPreflightResponseAPI {
  ok: boolean;
  missing_credential: boolean;
  invalid_credential: boolean;
  missing_agent_keys: string[];
  disabled_agent_keys: string[];
  errors: string[];
}

export function fetchWorkspaceModelConfig(
  workspaceSlug: string,
  signal?: AbortSignal,
): Promise<WorkspaceModelConfigAPI> {
  return api.get<WorkspaceModelConfigAPI>(
    `/api/v1/workspaces/${encodeURIComponent(workspaceSlug)}/model-config`,
    workspaceQueryParams(),
    signal,
  );
}
export function upsertOpenRouterKey(
  workspaceSlug: string,
  body: OpenRouterKeyUpsertRequestAPI,
): Promise<CredentialStatusAPI> {
  return api.put<CredentialStatusAPI>(
    `/api/v1/workspaces/${encodeURIComponent(workspaceSlug)}/model-config/openrouter-key`,
    body,
  );
}

export function deleteOpenRouterKey(
  workspaceSlug: string,
): Promise<DeleteOpenRouterKeyResponseAPI> {
  return api.del<DeleteOpenRouterKeyResponseAPI>(
    `/api/v1/workspaces/${encodeURIComponent(workspaceSlug)}/model-config/openrouter-key`,
    workspaceQueryParams(),
  );
}

export function testOpenRouterKey(
  workspaceSlug: string,
  body: OpenRouterKeyTestRequestAPI,
): Promise<CredentialTestResponseAPI> {
  return api.post<CredentialTestResponseAPI>(
    `/api/v1/workspaces/${encodeURIComponent(workspaceSlug)}/model-config/openrouter-key/test`,
    body,
    { params: workspaceQueryParams() },
  );
}

export function updateAgentModelConfig(
  workspaceSlug: string,
  agentKey: string,
  body: AgentModelConfigUpdateRequestAPI,
): Promise<AgentModelConfigAPI> {
  return api.patch<AgentModelConfigAPI>(
    `/api/v1/workspaces/${encodeURIComponent(workspaceSlug)}/model-config/agents/${encodeURIComponent(agentKey)}`,
    body,
    { params: workspaceQueryParams() },
  );
}

export function testAgentModelConfig(
  workspaceSlug: string,
  agentKey: string,
): Promise<AgentConfigTestResponseAPI> {
  return api.post<AgentConfigTestResponseAPI>(
    `/api/v1/workspaces/${encodeURIComponent(workspaceSlug)}/model-config/agents/${encodeURIComponent(agentKey)}/test`,
    {},
    { params: workspaceQueryParams() },
  );
}

export function preflightModelConfig(
  workspaceSlug: string,
  agentKeys: string[],
): Promise<ModelConfigPreflightResponseAPI> {
  return api.post<ModelConfigPreflightResponseAPI>(
    `/api/v1/workspaces/${encodeURIComponent(workspaceSlug)}/model-config/preflight`,
    { agent_keys: agentKeys },
    { params: workspaceQueryParams() },
  );
}
